Sprint 12

YanikMax opened the pull request.

Commit:
Свёрстан Launch Screen по макету;
Свёрстан главный экран с лентой изображений.

Commit:
Добавлена градиентная подложка под дату.
Экран профиля сверстан с помощью кода.
Добавлен экран отдельного изображения с возможность шеринга картинки.

Commit:
Исправил констрейнты аватара.
Изменил цвет статус-бара на белый.
Полностью убран код связанный с градиентной подложкой.
Для ImageListCell @IBOutlet сделал приватными.
При импорте UIKit убрал Foundation.
Константу ShowSingleImageSegueIdentifier переименовал (не знаю что на меня нашло).
Извелечение опционала в ImagesListViewController реализовал через guard let.
При верстке кодом использовал предложенное флоу.

Commit:
Для входа в приложение пользователь авторизуется через OAuth2.0;
В проекте создан файл Constants.swift, в нём есть константы AccessKey, SecretKey, RedirectURI и AccessScope;
Свёрстан контроллер WebView;
Реализован KVO — приложение показывает, как обновляется прогресс загрузки веб-страницы;
Проверка необходимости авторизации выделена в отдельный контроллер SplashViewController;
Добавлена структура для декодинга JSON-ответа от POST-запроса авторизации Unsplash — OAuthTokenResponseBody;
Создан новый класс OAuth2Service.

Commit:
Переименовал константы, которые были написаны с заглавной буквы;
Исправил в сториборде работу кнопки назад;
В WebViewViewController организовал код согласно общепринятым нормам;
Убрал force unwrapping и заменил на guard
Добавил trailingAnchor, если будет слишком длинное имя;
Добавил константы из файла Constants;
Исправил ключ AccessKey (в начале был лишний символ @)
Приложение было зарегистрировано, но слетели галочки разрешений (исправил).

Commit:
Устранено состояние гонки в OAuthService.
Устранено состояние гонки c помощью блокировки UI.
Сделан сетевой слой для экрана профиля.
Пока идёт загрузка данных профиля, показывается Splash Screen.
Для отображения аватарки пользователя использована библиотека Kingfisher.
В проект добавлен SwiftKeychainWrapper.

Commit:
Добавил использование progressView.setProgress для более плавного заполнения прогресс-бара;
Исправлена опечатка в методе didTapBackButton;
Исправлены названия констант в файле констант;
Перенастроил дата форматтер по макету;
Переименовал константу profile_image в файле ProfileImageService.swift в стиле camelCase;
Переименовал константу DidChangeNotification в файле ProfileImageService.swift в стиле camelCase;
Добавил настройки для аватара, чтобы он был круглым;
Добавил настройки для descriptionLabel, чтобы текст был многострочным и не обрезался;
Вынес значения констант из файла OAuth2Service.swift;
Переписал операцию без force unwrap в файле SplashViewController.swift;
Исправил поток авторизации согласно комментарию;
Переписал операцию без force unwrap в файле ProfileService.swift;
Исправил ссылку для получения профиля в файле ProfileService.swift;
Поля lastName и bio сделал опциональным типом;
Пункт 4.2 в сервис ProfileService не добавлена функция fetchProfile, которая делает запрос GET /me с Bearer Token (но данная функция у меня присутствует в коде).

Commit:
Спасибо Вам, Лилия, за расширенные комментарии.
Сохранение токена перенес в SplashViewController;
Сохранение переписал в SwiftKeychainWrapper;
Изменил путь апи профиля и картинки;
Заменил использование метода на showAuthenticationScreen;
Изменил ключи;
Фон макета изменил на YPBlack;
Констрейнты ширины и высоты убрал;
ProfileService сделано синглтоном.

Commit:
Создан сетевой слой для экрана с лентой фотографий;
Реализована функциональность лайков;
Добавлена возможность полноэкранного просмотра фотографий;
Добавлена возможность выхода из аккаунта.

andrew-kuzn-yndx commented on Image Feed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,125 @@
+        if
+            let url = navigationAction.request.url,
+            let urlComponents = URLComponents(string: url.absoluteString),
+            urlComponents.path == "/oauth/authorize/native",
```
Этот путь тоже можно хранить в Constants, не зря ведь ты его делал)

andrew-kuzn-yndx commented on Image Feed/ImagesList/ImagesListCell.swift:
```diff
@@ -0,0 +1,60 @@
+        cellImage.kf.indicatorType = .activity
+        
+        let placeholderImage = UIImage(named: "placeholderImage")
+        cellImage.kf.setImage(with: url, placeholder: placeholderImage) { result in
```
😱 Забыл [weak self]

andrew-kuzn-yndx commented on Image Feed/ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,188 @@
+                    if self.lastLoadedPage == nil {
+                        self.lastLoadedPage = 1
+                    } else {
+                        self.lastLoadedPage! += 1
```
Чтобы немного улучшить синтаксис, можно использовать тернарный оператор

andrew-kuzn-yndx commented on Image Feed/ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,188 @@
+                    NotificationCenter.default.post(name: ImagesListService.didChangeNotification, object: nil)
+                    
+                case .failure(let error):
+                    print(error.localizedDescription)
```
Ошибку стоит обработать, показать алерт, например

andrew-kuzn-yndx commented on Image Feed/ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,188 @@
+                        completion(.success(()))
+                    }
+                case .failure(let error):
+                    assertionFailure("like error: \(error)")
```
И тут)

andrew-kuzn-yndx commented on Image Feed/ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,188 @@
+    }
+    
+    private func makeRequest(authToken: String, page: Int) -> URLRequest? {
+        guard var urlComponents = URLComponents(string: "https://api.unsplash.com/photos") else {
```
Это тоже можно поместить в Constants

andrew-kuzn-yndx commented on Image Feed/ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,188 @@
+    }
+    
+    private func likeRequest(photoId: String, isLike: Bool) -> URLRequest? {
+        let baseURL = "https://api.unsplash.com"
```
И это

andrew-kuzn-yndx commented on Image Feed/ImagesList/ImagesListViewController.swift:
```diff
@@ -0,0 +1,155 @@
+        formatter.dateFormat = "dd MMMM yyyy"
+        formatter.locale = Locale(identifier: "ru_RU")
+        return formatter
+    }()
```
dateFormatter лучше создавать в отдельном extension. Он может пригодиться в разных местах приложения, и поэтому лучше создать его один раз и в одном месте)

andrew-kuzn-yndx commented on Image Feed/ImagesList/ImagesListViewController.swift:
```diff
@@ -0,0 +1,155 @@
+extension ImagesListViewController{
+    private func notifications() {
+        imagesListServiceObserver = NotificationCenter.default
+            .addObserver(
```
Сейчас не принципиально, но в будущем всегда учитывай, что при добавлении Observer, нужно предусматривать и его удаление, чтобы не создать утечку памяти

andrew-kuzn-yndx commented on Image Feed/ImagesList/ImagesListViewController.swift:
```diff
@@ -0,0 +1,155 @@
+                UIBlockingProgressHUD.dismiss()
+                
+            case .failure(let error):
+                UIBlockingProgressHUD.dismiss()
```
Дублирование UIBlockingProgressHUD. Его можно вывести за пределы switch, так как его вызывать в любом случае)

andrew-kuzn-yndx commented on Image Feed/ImagesList/ImagesListViewController.swift:
```diff
@@ -0,0 +1,155 @@
+                let okAction = UIAlertAction(title: "Ок", style: .default)
+                
+                alert.addAction(okAction)
+                self.present(alert, animated: true)
```
Для порядка в структуре, создание и показ алерта лучше вывести в отдельную функцию

andrew-kuzn-yndx commented on Image Feed/Profile/ProfileService.swift:
```diff
@@ -0,0 +1,86 @@
+        }
+        
+        // создание GET запроса с заголовком Authorization
+        guard let url = URL(string: "https://api.unsplash.com/me") else {
```
url можно в Constants

andrew-kuzn-yndx commented on Image Feed/Profile/ProfileViewController.swift:
```diff
@@ -0,0 +1,199 @@
+        fetchProfileAndUpdateUI()
+        
+        profileImageServiceObserver = NotificationCenter.default
+            .addObserver(
```
Здесь тоже нужно подумать о удалении наблюдателя

andrew-kuzn-yndx commented on Image Feed/SingleImageView/SingleImageViewController.swift:
```diff
@@ -0,0 +1,133 @@
+        dismiss(animated: true, completion: nil)
+    }
+    
+    @IBAction func didTapShareButton(_ sender: UIButton) {
```
Подзабыл private

andrew-kuzn-yndx commented on Image Feed/SplashViewController.swift:
```diff
@@ -0,0 +1,145 @@
+    }
+    
+    private func fetchProfileImage(username: String) {
+        ProfileImageService.shared.fetchProfileImageURL(username: username) { _ in
```
Забыл [weak self]

Commit:
Добавлены [weak self] в ImageListCell в метод setupCell и в SplashViewController в метод fetchProfileImage;
В ImageListService использован тернарный оператор заместо конструкции if-else;
В ImageListService в методы fetchPhotosNextPage и  changeLike добавлены алерты для обработки ошибки;
URL были перенесены в файл Constants;
dateFormatter был перенесен в отдельный extension;
При добавлении наблюдателя Observer так же была добавлена реализация его удаления;
UIBlockingProgressHUD вынесено за пределы switch;
@IBAction func didTapShareButton добавлен private.
